From ba789c682c5b419eb860a8729c73704a4c1dc143 Mon Sep 17 00:00:00 2001
From: verma <amit.verma@amd.com>
Date: Mon, 31 Jul 2023 16:46:16 +0530
Subject: [PATCH 1/4] media: i2c: Add max9296 and ox03f10 in Makefile

Signed-off-by: verma <amit.verma@amd.com>
---
 drivers/media/i2c/Makefile | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/media/i2c/Makefile b/drivers/media/i2c/Makefile
index d39c591..43d0657 100644
--- a/drivers/media/i2c/Makefile
+++ b/drivers/media/i2c/Makefile
@@ -58,6 +58,7 @@ obj-$(CONFIG_VIDEO_M52790) += m52790.o
 obj-$(CONFIG_VIDEO_M5MOLS) += m5mols/
 obj-$(CONFIG_VIDEO_MAX9271_LIB) += max9271.o
 obj-$(CONFIG_VIDEO_MAX9286) += max9286.o
+obj-$(CONFIG_VIDEO_MAX9296) += max9296.o
 obj-$(CONFIG_VIDEO_ML86V7667) += ml86v7667.o
 obj-$(CONFIG_VIDEO_MSP3400) += msp3400.o
 obj-$(CONFIG_VIDEO_MT9M001) += mt9m001.o
@@ -100,6 +101,7 @@ obj-$(CONFIG_VIDEO_OV9282) += ov9282.o
 obj-$(CONFIG_VIDEO_OV9640) += ov9640.o
 obj-$(CONFIG_VIDEO_OV9650) += ov9650.o
 obj-$(CONFIG_VIDEO_OV9734) += ov9734.o
+obj-$(CONFIG_VIDEO_OX03F10) += ox03f10.o
 obj-$(CONFIG_VIDEO_RDACM20) += rdacm20.o
 obj-$(CONFIG_VIDEO_RDACM21) += rdacm21.o
 obj-$(CONFIG_VIDEO_RJ54N1) += rj54n1cb0c.o
-- 
1.8.3.1

