From 9b511e3789a0dc75c21a256ad2773d673318ef97 Mon Sep 17 00:00:00 2001
From: verma <amit.verma@amd.com>
Date: Mon, 31 Jul 2023 16:49:04 +0530
Subject: [PATCH 2/4] media: i2c: add max9296 and ox03f10 in Kconfig

Signed-off-by: verma <amit.verma@amd.com>
---
 drivers/media/i2c/Kconfig | 25 +++++++++++++++++++++++++
 1 file changed, 25 insertions(+)

diff --git a/drivers/media/i2c/Kconfig b/drivers/media/i2c/Kconfig
index c684888..39fb59b 100644
--- a/drivers/media/i2c/Kconfig
+++ b/drivers/media/i2c/Kconfig
@@ -688,6 +688,19 @@ config VIDEO_OV9734
 	  To compile this driver as a module, choose M here: the
 	  module's name is ov9734.
 
+config VIDEO_OX03F10
+	tristate "Omni OX03F10 sensor support"
+	depends on I2C && VIDEO_DEV
+	select MEDIA_CONTROLLER
+	select VIDEO_V4L2_SUBDEV_API
+	select V4L2_FWNODE
+	help
+	  This is a Video4Linux2 sensor driver for the Omni
+	  OX03F10 camera.
+
+	  To compile this driver as a module, choose M here: the
+	  module will be called ox03f10.
+
 config VIDEO_RDACM20
 	tristate "IMI RDACM20 camera support"
 	depends on I2C
@@ -1235,6 +1248,18 @@ config VIDEO_MAX9286
 	  To compile this driver as a module, choose M here: the
 	  module will be called max9286.
 
+config VIDEO_MAX9296
+	tristate "Maxim MAX9296 GMSL deserializer support"
+	depends on I2C && I2C_MUX
+	depends on OF_GPIO
+	select V4L2_FWNODE
+	select VIDEO_V4L2_SUBDEV_API
+	select MEDIA_CONTROLLER
+	help
+	  This driver supports the Maxim MAX9296 GMSL deserializer.
+	  To compile this driver as a module, choose M here: the
+	  module will be called max9296.
+
 config VIDEO_ML86V7667
 	tristate "OKI ML86V7667 video decoder"
 	depends on VIDEO_DEV && I2C
-- 
1.8.3.1

